Enable set sequence nr

YLGH opened the pull request:
Summary:
In some cases (especially those involving collective calls) - we would want to always kick off a collective call first before running going down another path.

For  example:

```
tbe lookup -> a2a ->
                     overarch
dense ------------->
```

if the forward code is written as
a2a_out = a2a
dense = dense_net
out = overarch(a2a_out, dense)
out.backward()

The current default is running backwards in the opposite order the forward is called. However, there is no data dependency between a2a and dense, so in reality either of them could be run first. We would like the a2a to run first because it provides optimal (on average) overlap.

Changing the seq_nr of a2a_out to something large enough would allow autograd engine to kick it off first.

Test Plan: Tests incoming

Differential Revision: D51445261

ezyang reviewed (approved):
this LGTM, not sure if Alban you want to have the final word here though, if so just bounce my review

albanD commented on test/test_set_sequence_nr.py:
```diff
@@ -0,0 +1,52 @@
+# Owner(s): ["module: autograd"]
```
Please move this to `test/autograd/test_seq_nr.py` and import it at the bottom of `test/test_autograd.py`.
Feel free to just add it to `test/test_autograd.py` given that it is very small.

albanD commented on test/run_test.py:
```diff
@@ -324,6 +324,7 @@ def skip_test_p(name: str) -> bool:
 CI_SERIAL_LIST = [
     "test_nn",
     "test_fake_tensor",
+    "test_set_sequence_nr",
```
Why does this need to be in the serial list? It should be ok to run it anytime?

YLGH replied:
oh whoops I mostly had no idea how to write a test and just templated off of test_fake_tensor

albanD commented on torch/csrc/autograd/python_cpp_function.cpp:
```diff
@@ -201,6 +201,13 @@ PyObject* THPCppFunction_sequence_nr(PyObject* self, PyObject* noargs) {
 }
+
+PyObject* THPCppFunction_set_sequence_nr(PyObject* self, PyObject* sequence_nr) {
+  auto& fn = *((THPCppFunction*)self)->cdata;
+  fn.set_sequence_nr(THPUtils_unpackUInt64(sequence_nr));
+  Py_RETURN_NONE;
```
```suggestion
  HANDLE_TH_ERRORS
  auto& fn = *((THPCppFunction*)self)->cdata;
  fn.set_sequence_nr(THPUtils_unpackUInt64(sequence_nr));
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
```

YLGH replied:
I didn't see any of the other methods in cpp_function having this (but I did have it in python_function.cpp)

albanD replied:
because most of them cannot raise c++ errors and handle python errors by hand (for historical reasons, these macro might not have existed when this code was written).
THPUtils_unpackUInt64 will throw a nice c++ error on error which you need to handle here.

YLGH replied:
ahh gotcha, thanks

albanD reviewed (commented):
Small details but mostly good.

albanD commented on test/test_autograd.py:
```diff
@@ -1,54 +1,83 @@
 # Owner(s): ["module: autograd"]
 
+import collections
```
I guess all the import shuffling is auto-formatting?
We usually don't do these unless they're from the lint to avoid shuffling back and forth depending on dev environments.

YLGH replied:
ah whoops - i'll let lint runner check to see if this is ok, otherwise i'll make it original format

YLGH replied:
reverted formatting changes

albanD reviewed (approved):
SGTM

YLGH committed:
Enable set sequence nr (#114120)

Summary:

In some cases (especially those involving collectiv calls) - we would want to always kick off a collective call first before running going down another path.

For  example:

```
tbe lookup -> a2a ->
                     overarch
dense ------------->
```

if the forward code is written as
a2a_out = a2a
dense = dense_net
out = overarch(a2a_out, dense)
out.backward()

The current default is running backwards in the opposite order the forward is called. However, there is no data dependency between a2a and dense, so in reality either of them could be run first. We would like the a2a to run first because it provides optimal (on average) overlap.

Changing the seq_nr of a2a_out to something large enough would allow autograd engine to kick it off first.

Test Plan: Tests incoming

Reviewed By: ezyang, albanD

Differential Revision: D51445261

YLGH commented:
@pytorchbot ,erge

YLGH commented:
@pytorchbot merge

YLGH commented:
@pytorchbot label "topic: not user facing"

YLGH commented:
@pytorchbot merge
